### ISSUE TITLE
Teach git bisect `--first-parent`

### ISSUE THREAD

**Issue by Denton-L** [gitgitgadget/git#439]
Bisect is a really useful feature when most of the commits are testable. However, in many GitHub users' workflow model, PRs are first-class while individual commits don't really matter that much so individual commits in a topic branch may not be testable but the first-parent of the master branch _would_ be.

Teach git bisect the `--first-parent` option to only bisect on first-parents.

This feature request has been discussed before [here](https://stackoverflow.com/questions/20240526/how-to-git-bisect-only-on-one-branchs-commits) and [here](https://public-inbox.org/git/CA+P7+xq-g3JPxKROa9Ne0EGiSgSqHKgqVG3eoW9Q-dP0yPOrVA@mail.gmail.com/) but it seems relatively obscure.

**Comment by dscho** [comment]
There also has been a PR at https://github.com/git/git/pull/484 as well as [a patch submission on the Git mailing list](https://public-inbox.org/git/20180828123234.44582-1-tiagonbotelho@hotmail.com/), which I am very disappointed to see abandoned.

Maybe a more active contributor (or at least a contributor who is more interested in receiving help to improve their patches) sees fit to pick it up and drive it to the finish line?

**Comment by dscho** [comment]
>  [a patch submission on the Git mailing list](https://public-inbox.org/git/20180828123234.44582-1-tiagonbotelho@hotmail.com/)

I just confirmed via [a quick search](https://public-inbox.org/git/?q=first-parent+bisect) that this was the latest iteration sent.

All of the offered concerns still hold true:

- The added regression test totally mismatches the existing tests, and refuses to use any part of the `setup` test case.
- Some parts of the patch are woefully underexplained in the commit message, namely
   ```
   @@ -329,6 +334,11 @@ static struct commit_list *do_find_bisection(struct commit_list *list,
    			if (0 <= weight(p))
    				continue;
    			for (q = p->item->parents; q; q = q->next) {
   +				if ((bisect_flags & BISECT_FIRST_PARENT)) {
   +					if (weight(q) < 0)
   +						q = NULL;
   +					break;
   +				}
    				if (q->item->object.flags & UNINTERESTING)
    					continue;
    				if (0 <= weight(q))
   ```
- The patch seems to change the logic [in an undesired way](https://public-inbox.org/git/xmqqlg8qh22q.fsf@gitster-ct.c.googlers.com/).

It is really sad that this patch was not developed further, it feels as if it was abandoned _so_ close to the goal.

**Comment by workingjubilee** [comment]
No promises! But I'll take a look at this. It's kind of fun to do this cleanup work.

**Comment by To1ne** [comment]
cc @tiagonbotelho

**Comment by dscho** [comment]
@To1ne Tiago is the person who opened https://github.com/git/git/pull/484 (and indicated that they are no longer interested in pursuing it, or no longer have time, or both).

**Comment by workingjubilee** [comment]
It required several updates just to bring in the recommended fixes and make the code conform with new changes in the project (since, well, it was a year), but now I am currently in the process of documenting my iteration on the patch.

**Comment by workingjubilee** [comment]
Success! It is a good day to ~~die~~ get constructive criticism!  https://public-inbox.org/git/20191105052141.15913-1-workingjubilee@gmail.com/T/#mffc30f9c2160311caed122ba4132a76a8da1f6e7

**Comment by Denton-L** [comment]
Closed by 47f0f94bc796037c43000a9852cdf9209a3c3274.
